src/lib.rs

```diff
@@ -57,18 +65,27 @@ pub mod ahp;
 pub use ahp::AHPForR1CS;
 use ahp::EvaluationsProvider;
 
+use crate::ahp::prover::ProverMsg;
+
 #[cfg(test)]
 mod test;
 
-/// The compiled argument system.
-pub struct Marlin<F: PrimeField, PC: PolynomialCommitment<F, DensePolynomial<F>>, FS: FiatShamirRng>(
+/// The compiled argument system.FiatShamiRng
+pub struct Marlin<
+    F: PrimeField,
+    PC: PolynomialCommitment<F, DensePolynomial<F>, S>,
+    S: CryptographicSponge,
+>(
     #[doc(hidden)] PhantomData<F>,
     #[doc(hidden)] PhantomData<PC>,
-    #[doc(hidden)] PhantomData<FS>,
+    #[doc(hidden)] PhantomData<S>,
 );
 
-impl<F: PrimeField, PC: PolynomialCommitment<F, DensePolynomial<F>>, FS: FiatShamirRng>
-    Marlin<F, PC, FS>
+impl<
+        F: PrimeField + Absorb,
```

Comment:
The addition of the Absorb trait bound is very problematic. Although it will improve efficiency of absorption, it makes the implementation far less generic. I am not certain how to improve this, considering that specialization of traits is an unstable feature.

Reply:
I resolved this by implementing the `fast_prove` and `fast_verify` methods that may be invoked when the PrimeField also satisfies the Absorb trait bound. I think this is a good trade off, since Absorb is not nearly as well implemented as Canonical(De)Serialize and it may be a pain for people to upgrade to the new implementation. Over time, I think the old method can be deprecated in favor of the new, but I think a prerequisite is implementing a derive Absorb macro.